gpio/zynq7000-gpio/gpio.c

```diff
@@ -0,0 +1,296 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Zynq-7000 GPIO controller
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Lukasz Kosinski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdlib.h>
+
+#include <sys/mman.h>
+#include <sys/platform.h>
+
+#include <phoenix/arch/zynq7000.h>
+#include <board_config.h>
+
+#include "gpio.h"
+
+
+/* GPIO registers */
+enum {
+	dlsw0 = 0, dmsw0, dlsw1, dmsw1,
+	data0 = 16, data1,
+	rodata0 = 24, rodata1,
+	dirm0 = 129, oen0, imr0, ier0, idr0, isr0, ityp0, ipol0, iany0,
+	dirm1 = 145, oen1, imr1, ier1, idr1, isr1, ityp1, ipol1, iany1,
+};
+
+
+/* GPIO pins */
+const int gpioPins[GPIO_BANKS][GPIO_PINS] = {
+	{
+		GPIO0_0, GPIO0_1, GPIO0_2, GPIO0_3,
+		GPIO0_4, GPIO0_5, GPIO0_6, GPIO0_7,
+		GPIO0_8, GPIO0_9, GPIO0_10, GPIO0_11,
+		GPIO0_12, GPIO0_13, GPIO0_14, GPIO0_15,
+		GPIO0_16, GPIO0_17, GPIO0_18, GPIO0_19,
+		GPIO0_20, GPIO0_21, GPIO0_22, GPIO0_23,
+		GPIO0_24, GPIO0_25, GPIO0_26, GPIO0_27,
+		GPIO0_28, GPIO0_29, GPIO0_30, GPIO0_31
+	},
+	{
+		GPIO1_0, GPIO1_1, GPIO1_2, GPIO1_3,
+		GPIO1_4, GPIO1_5, GPIO1_6, GPIO1_7,
+		GPIO1_8, GPIO1_9, GPIO1_10, GPIO1_11,
+		GPIO1_12, GPIO1_13, GPIO1_14, GPIO1_15,
+		GPIO1_16, GPIO1_17, GPIO1_18, GPIO1_19,
+		GPIO1_20, GPIO1_21, -1, -1,
+		-1, -1, -1, -1,
+		-1, -1, -1, -1,
+	},
+};
+
+
+static struct {
+	volatile uint32_t *base; /* GPIO registers base address */
+} gpio_common;
+
+
+static int gpio_checkPin(unsigned int bank, unsigned int pin)
+{
+	if ((bank >= GPIO_BANKS) || (pin >= GPIO_PINS)) {
+		return -EINVAL;
+	}
+
+	if (gpioPins[bank][pin] < 0) {
+		return -EINVAL;
+	}
+
+	return EOK;
+}
+
+
+int gpio_readPort(unsigned int bank, uint32_t *val)
+{
+	if (bank == 0) {
+		*val = *(gpio_common.base + rodata0);
+	}
+	else if (bank == 1) {
+		*val = *(gpio_common.base + rodata1);
+	}
+	else {
+		return -EINVAL;
+	}
+
+	return EOK;
+}
+
+
+int gpio_writePort(unsigned int bank, unsigned char word, uint32_t maskval)
```

Comment:
Why not write both halves in this function? Also, there is no way to modify only some bits - why not pass mask and val?

Reply:
I wanted to keep the operation 'atomic' - single write to GPIO register. Since it's not that important I changed the code as suggested.